### PR TITLE
Implement `Relude.Extra.Types` module

### LINKED ISSUE #156: Implement common type-families in the `Relude.Extra.Types` module
Sometimes they are needed for different tasks. Maybe we can put some of the into our `relude` to ease various tasks. I can think of the following useful type families:

- [ ] `type family (++) (xs :: [k]) (ys :: [k]) :: [k]`
- [ ] `type family All (f :: k -> Constraint) (xs :: [k]) :: Constraint`
- [ ] `Fst/Snd` to get first second elements of different tuples

This is something I can think of now. Maybe more useful functions/operators can be added later.

### PR DESCRIPTION
**Pull request by TheMatten**
Resolves #156.

Additionally to `type (++)` it adds `AllHave`, `Fst` and `Snd` - should they be kept or removed for now?

### HLint

- [ ] I've changed the exposed interface (add new reexports, remove reexports, rename reexported things, etc.).
  - [ ] I've updated [`hlint.dhall`](https://github.com/kowainik/relude/blob/master/hlint/hlint.dhall) accordingly to my changes (add new rules for the new imports, remove old ones, when they are outdated, etc.).
  - [ ] I've generated the new `.hlint.yaml` file (see [this instructions](https://github.com/kowainik/relude#generating-hlintyaml)).

### General

- [ ] I've updated the [CHANGELOG](https://github.com/kowainik/relude/blob/master/CHANGELOG.md) with the short description of my latest changes.
- [x] All new and existing tests pass.
- [x] I keep the code style used in the files I've changed (see [style-guide](https://github.com/kowainik/org/blob/master/style-guide.md#haskell-style-guide) for more details).
- [x] I've used the [`stylish-haskell` file](https://github.com/kowainik/relude/blob/master/.stylish-haskell.yaml).
- [x] My change requires the documentation updates.
  - [x] I've updated the documentation accordingly.
- [ ] I've added the `[ci skip]` text to the docs-only related commit's name.

### CODE REVIEW

File: relude.cabal [hunk 1]
```diff
@@ -142,6 +142,7 @@ library
                            Relude.Extra.Newtype
                            Relude.Extra.Tuple
                            Relude.Extra.Type
+                           Relude.Extra.Types
```
Comment (chshersh) [inline relude.cabal]: I think that there is no need to introduce new `Relude.Extra.Types` module. Let's just put all these functions into `Relude.Extra.Type`

Comment (TheMatten) [reply relude.cabal]: Sorry, I took title in issue literally :smile:...

File: src/Relude/Extra/Type.hs [hunk 2]
```diff
@@ -1,18 +1,27 @@
...
+{-# LANGUAGE DataKinds           #-}
+{-# LANGUAGE PolyKinds           #-}
 {-# LANGUAGE ScopedTypeVariables #-}
 {-# LANGUAGE TypeApplications    #-}
+{-# LANGUAGE TypeFamilies        #-}
+{-# LANGUAGE TypeOperators       #-}
 
 {- |
 Copyright:  (c) 2018-2019 Kowainik
+                2019 TheMatten
```
Comment (chshersh) [inline src/Relude/Extra/Type.hs]: The code itself is good. But I need to ask you to not add your personal copyright notice to this file so Kowainik can keep ownership of all code. When everything is licensed under `Kowainik` name, then it's much easier to resolve various legal issues that can arise in future. For example, if we keep this copyright notice and decide to change the license in future, we will need to ask your personal agreement to do that. This seems like a high price to pay for implementing our proposal to add several type-level functions :slightly_smiling_face:

Comment (TheMatten) [reply src/Relude/Extra/Type.hs]: Wasn't really sure about this and hoped you will sort that out - I don't care about these functions at all :D

Comment (TheMatten) [reply src/Relude/Extra/Type.hs]: Done - will probably ask instead in the future when not sure :slightly_smiling_face:

Comment (chshersh) [reply src/Relude/Extra/Type.hs]: Thanks for understanding! And no worries 🙂

**Pull request merged**